### PR TITLE
chore(setup): initial data structure

### PR DESCRIPTION
**Pull request by larissagaulia**

### CODE REVIEW

File: lib/spike-detector.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
+var moment = require('moment');
+
+var normalTraffic = {};
```
Comment (vladikoff) [inline lib/spike-detector.js]: this should be scoped to `new SpikeD(config);` object. 

so below you should reference `this.normalTraffic`

File: lib/spike-detector.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  var day = now.format('YYMMDD');
+  var hour = now.format('H');
+
+  //if it's a new day
+    //TODO: delete old data, meaning entry now.subtract((config.days+1), 'days')
+    //create today's entry
+    //this.updateNormalTraffic();
+  if(!normalTraffic[day]){
+  }
+  _incrementTo(day, hour);
```
Comment (vladikoff) [inline lib/spike-detector.js]: this should call `this._incrementTo()`

File: lib/spike-detector.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
+var moment = require('moment');
+
+var normalTraffic = {};
+
+function SpikeD(config) {
```
Comment (vladikoff) [inline lib/spike-detector.js]: nit: i'd call this `options` instead of `config`.

File: lib/spike-detector.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
+var moment = require('moment');
+var _ = require('underscore');
```
Comment (vladikoff) [inline lib/spike-detector.js]: let's use Lodash

File: package.json [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    "grunt": "^1.0.1",
+    "grunt-eslint": "^19.0.0",
+    "load-grunt-tasks": "^3.5.2",
+    "mocha": "^3.0.2",
+    "sinon": "^1.17.5"
+  },
+  "dependencies": {
+    "debug": "^2.2.0",
+    "moment": "^2.14.1",
+    "underscore": "^1.8.3"
```
Comment (vladikoff) [inline package.json]: https://github.com/larissagaulia/spike-detection/pull/1/files#r78591415

File: lib/spike-detector.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+var debug = require('debug')('spikeD');
+var events = require('events');
+var eventEmitter;
+
+module.exports = function (options) {
+
+  function SpikeD() {
+    eventEmitter = new events.EventEmitter();
+    eventEmitter.on('checkForSpike', this.checkForSpike.bind(this));
```
Comment (vladikoff) [inline lib/spike-detector.js]: looks like this doesn't belong here. we should use `.emit` and emit the event on one of the `increments` once we detect the spike. the `.on` should be in 'customs'

File: lib/spike-detector.js (hunk 4 above)
Comment (vladikoff) [reply lib/spike-detector.js]: https://www.npmjs.com/package/lodash

File: package.json [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    "url": "https://github.com/larissagaulia/spike-detection/issues"
+  },
+  "homepage": "https://github.com/larissagaulia/spike-detection",
+  "devDependencies": {
+    "chai": "^3.5.0",
+    "eslint-config-fxa": "^2.1.0",
+    "grunt": "^1.0.1",
+    "grunt-eslint": "^19.0.0",
+    "load-grunt-tasks": "^3.5.2",
+    "mocha": "^3.0.2",
```
Comment (vladikoff) [inline package.json:37]: coverage: http://ricostacruz.com/til/mocha-instanbul-coverage

File: lib/deviation.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
+var _ = require('lodash');
+
+function Deviation(locationData, userLocale) {
+  'use strict';
+
+  this.getAverage = function (array) {
+    var count = array.length;
```
Comment (vladikoff) [inline lib/deviation.js:7]: do we need a guard for array.length `0`?

File: lib/deviation.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    var variance =  _.reduce(array, function(sum, n) {
+      return sum + Math.pow((n - average), 2)/length;
+    }, 0)
+
+    return Math.sqrt(variance);
+  };
+
+}
+
+module.exports = Deviation;
```
Comment (vladikoff) [inline lib/deviation.js]: need a line break here

File: lib/spike-detector.js [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+    debug('last hour count was' + ' ' + this.normalTraffic[previousHourDate.format('YYMMDD')][previousHourDate.format('H')]);
+    return this.normalTraffic[previousHourDate.format('YYMMDD')][previousHourDate.format('H')];
+  };
+
+  //TODO: make "spike factor" configurable
+  SpikeD.prototype.checkForSpike = function (date) {
+    if(!moment.isMoment(date)){
+      date = moment(date);
+    }
+    var previousHour = date.subtract(1, 'hour').format('H');
```
Comment (vladikoff) [inline lib/spike-detector.js:33]: should this be `previous` or `current` hour? 

so if it is 3:30pm, is the current hour `3` ?

File: lib/spike-detector.js [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+    var previousData = [];
+    
+    var previousDay = undefined;
+    //look at same hours on previous days
+    for(var day=1; day <= options.days; day++){
+      previousDay = this.getNormalTrafficDay(date.subtract(1, 'days'));
+      if(previousDay && previousDay[previousHour]) {
+        previousData.push(previousDay[previousHour]);
+      } else break;
+    }
```
Comment (vladikoff) [inline lib/spike-detector.js:51]: does the code above need clean up `previousHourCount` vs `previousData`?

File: test/deviation.js [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+  
+  it('gets average', function () {
+  	assert.equal(deviation.getAverage(testArray), 15);
+  });
+
+  it('gets standart deviation', function () {
+  	assert.equal(deviation.getStandardDeviation(testArray), 7.0710678118654755);
+  });
+
+});
```
Comment (vladikoff) [inline test/deviation.js]: need a new line here

File: test/deviation.js [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+
+var chai = require('chai');
+var assert = chai.assert;
+var Deviation = require('../lib/deviation');
+
+describe('deviation', function () {
+  'use strict';
+
+  var deviation = new Deviation();
+  var testArray = [5, 10, 15, 20, 25];
```
Comment (vladikoff) [inline test/deviation.js:11]: need test with empty array

File: lib/spike-detector.js (hunk 10 above)
Comment (larissagaulia) [reply lib/spike-detector.js:33]: we only call checkForSpike in the first increment of a new hour, but the check occurs for the previous time slot. 
Ex: 4:03pm occurred the first request, increment() got called, _incrementTo checks that 4pm slot doesnt exist yet and initializes it. Then it checks if there was an anomaly in the 3pm slot (the previous hour).

File: lib/spike-detector.js (hunk 11 above)
Comment (larissagaulia) [reply lib/spike-detector.js:51]: hmm, I'm renaming the variables to see if it makes it easier to read

File: lib/spike-detector.js (hunk 10 above)
Comment (vladikoff) [reply lib/spike-detector.js:33]: > Then it checks if there was an anomaly in the 3pm slot (the previous hour).

what is your reasoning for that?

File: lib/spike-detector.js [hunk 14]
```diff
@@ -0,0 +1,110 @@
+var moment = require('moment');
+var _ = require('lodash');
+var debug = require('debug')('spikeD');
+
+var events = require('events');
+var eventEmitter;
```
Comment (vladikoff) [inline lib/spike-detector.js]: not used anymore i think?

File: lib/spike-detector.js (hunk 10 above)
Comment (larissagaulia) [reply lib/spike-detector.js:33]: I had to decide when to check... Every new hour seemed not bad. Do you think is a bad idea?

File: lib/spike-detector.js [hunk 15]
```diff
@@ -0,0 +1,115 @@
+var moment = require('moment');
+var _ = require('lodash');
+var debug = require('debug')('spikeD');
+
+var events = require('events');
+
+var Deviation = require('./deviation');
+var deviation;
```
Comment (vladikoff) [inline lib/spike-detector.js:8]: This probably should not be global outside of `function SpikeD() {`
